python/cudf/cudf/core/udf/pipeline.py

```diff
@@ -50,3 +66,168 @@ def wrapper(*args):
         return to_udf_table._apply(func)
 
     return wrapper
+
+
+def masked_arrty_from_np_type(dtype):
+    """
+    Return a type representing a tuple of arrays,
+    the first element an array of the numba type
+    corresponding to `dtype`, and the second an
+    array of bools represe
+    """
+    nb_scalar_ty = numpy_support.from_dtype(dtype)
+    return Tuple((nb_scalar_ty[::1], libcudf_bitmask_type[::1]))
+
+
+def construct_signature(df, return_type):
+    """
+    Build the signature of numba types that will be used to
+    actually JIT the kernel itself later, accounting for types
+    and offsets
+    """
+
+    # Tuple of arrays, first the output data array, then the mask
+    return_type = Tuple((return_type[::1], boolean[::1]))
+    offsets = []
+    sig = [return_type]
+    for col in df._data.values():
+        sig.append(masked_arrty_from_np_type(col.dtype))
+        offsets.append(int64)
+
+    # return_type + data,masks + offsets + size
+    sig = void(*(sig + offsets + [int64]))
+
+    return sig
+
+
+@cuda.jit(device=True)
+def mask_get(mask, pos):
+    return (mask[pos // mask_bitsize] >> (pos % mask_bitsize)) & 1
+
+
+kernel_template = """\
+def _kernel(retval, {input_columns}, {input_offsets}, size):
+    i = cuda.grid(1)
+    ret_data_arr, ret_mask_arr = retval
+    if i < size:
+{masked_input_initializers}
+        ret = {user_udf_call}
+        ret_data_arr[i] = {ret_value}
+        ret_mask_arr[i] = {ret_valid}
+"""
+
+unmasked_input_initializer_template = """\
+        d_{idx}, m_{idx} = input_col_{idx}
+        masked_{idx} = Masked(d_{idx}[i], True)
+"""
+
+masked_input_initializer_template = """\
+        d_{idx}, m_{idx} = input_col_{idx}
+        masked_{idx} = Masked(d_{idx}[i], mask_get(m_{idx}, i + offset_{idx}))
+"""
+
+
+@annotate("DEFINE", color="yellow", domain="cudf_python")
+def _define_function(df, scalar_return=False):
+    # Create argument list for kernel
+    input_columns = ", ".join(
+        [f"input_col_{i}" for i in range(len(df.columns))]
```

Comment:
`.columns` is expensive -- use` ._data` instead (bonus: works for `Series` as well)

Reply:
Fixed

Reply:
Does that mean that this code is now dead? Or is an update here needed too?